lms/static/scripts/frontend_apps/components/GradingComment.tsx

```diff
@@ -0,0 +1,141 @@
+import type { TextareaProps } from '@hypothesis/frontend-shared';
+import {
+  Button,
+  CancelIcon,
+  CloseButton,
+  Dialog,
+  NoteFilledIcon,
+  NoteIcon,
+  PointerUpIcon,
+  Textarea,
+} from '@hypothesis/frontend-shared';
+import classnames from 'classnames';
+import { useCallback, useId, useRef, useState } from 'preact/hooks';
+
+type CommentPopoverProps = {
+  disabled: boolean;
+  comment: string;
+  onInput?: TextareaProps['onInput'];
+  onSubmit?: (e: Event) => Promise<boolean>;
+  closePopover: () => void;
+};
+
+function CommentPopover({
+  disabled,
+  comment,
+  onInput,
+  onSubmit,
+  closePopover,
+}: CommentPopoverProps) {
+  const commentRef = useRef<HTMLTextAreaElement | null>(null);
+  const commentId = useId();
+
+  return (
+    <Dialog
+      variant="custom"
+      title=""
+      classes={classnames(
+        'w-80 p-3',
+        'shadow border rounded bg-white',
+        'absolute top-[calc(100%+3px)] right-0'
+      )}
+      data-testid="comment-popover"
+      onClose={closePopover}
+      initialFocus={commentRef}
+      restoreFocus
+      closeOnClickAway
+      closeOnEscape
+    >
+      <PointerUpIcon
+        className={classnames(
+          'text-grey-3 fill-white',
+          'absolute inline z-2 w-[15px]',
+          // Position arrow over "Add comment" button
+          'right-[7px] top-[-9px]'
+        )}
+      />
+      <div className="flex items-center">
+        <label htmlFor={commentId} className="font-bold">
+          Add a comment:
+        </label>
+        <div className="grow" />
+        <CloseButton
+          title="Close comment"
+          classes="hover:bg-grey-3/50"
+          data-testid="comment-textless-close-button"
+        />
+      </div>
+      <Textarea
+        id={commentId}
+        classes="mt-1"
+        rows={10}
+        value={comment}
+        onInput={onInput}
+        elementRef={commentRef}
+      />
+      <div className="flex flex-row-reverse space-x-2 space-x-reverse mt-3">
+        <Button
+          variant="primary"
+          disabled={disabled}
+          onClick={async e => {
+            if (await onSubmit?.(e)) {
+              closePopover();
+            }
+          }}
+          data-testid="comment-submit-button"
+        >
+          Submit Grade
+        </Button>
+        <Button
+          icon={CancelIcon}
+          onClick={closePopover}
+          data-testid="comment-close-button"
+        >
+          Close
+        </Button>
+      </div>
+    </Dialog>
+  );
+}
+
+export type GradingCommentProps = Omit<CommentPopoverProps, 'closePopover'>;
+
+export default function GradingComment({
+  disabled,
+  comment,
+  onInput,
+  onSubmit,
+}: GradingCommentProps) {
+  const [showCommentPopover, setShowCommentPopover] = useState(false);
+  const closeCommentPopover = useCallback(
+    () => setShowCommentPopover(false),
+    []
+  );
+  const commentIsSet = !disabled && !!comment;
```

Comment:
Is there a reason that `commentIsSet` isn't defined in terms of the comment alone? If the comment is non-empty but the user is prevented from editing it for some reason (permissions?), the comment is still present.

Reply:
The reason to compute it from both "disabled" and "comment" is to take loading state into consideration (`disabled` is `true` while loading a student), otherwise the icon would not change from filled to not-filled or viceversa, until we load the new student.

But perhaps this is actually the desired behavior. If two consecutive students have a comment, current approach would make the icon switch from filled to not-filled until the new student is loaded, when it would change back to filled.

I will change it to not take `disabled` into consideration.

Reply:
> But perhaps this is actually the desired behavior. If two consecutive students have a comment, current approach would make the icon switch from filled to not-filled until the new student is loaded, when it would change back to filled.

Ah OK, so the `disabled` state is also being used for a `loading` state. I think the current behavior is reasonable then, its just that the dual use of `disabled` is not obvious. An explicit `loading` state could make things clearer.

Reply:
Yep, I was in fact thinking that while I was writing the comment 😅

In fact I was even going to suggest to use a third icon (loading spinner?) for the loading state. I'll see if it looks too redundant with the existing one.

Reply:
Too many spinners (maybe even just two) can feel distracting and cheap IMO. I would opt for a static indicator such as an icon variant (which could just be a contrast adjustment to one of the existing icons) if anything.

Reply:
The I will just stick to a transition to the non-filled icon as before, but based on a new `loading` prop instead of `disabled`.